### DISCUSSION TITLE
Problem with DID Auth over CHAPI

### DISCUSSION THREAD

**Discussion by leonfhl** [learningeconomy/LearnCard#161] [Help]
Hi all, we are trying to implement DID Auth in our own UI, using storeCredentialViaChapiDidAuth(unsignedCredential) to issue to **LearnCard** works, but when same method is used for **Veres Wallet**  it fails with:
```
{  success: false,
  reason: 'auth failed verification' }
```
According to the [LearnEconomy SDK docs](https://docs.learncard.com/learn-card-sdk/learncard-core/chapi/using-learncard-to-interact-with-a-chapi-wallet) this means: **"the user's wallet software failed the verification challenge"**. I suspect is related with _acceptedCryptosuites_ o _acceptedMethod_ but didn't know how to debug it and what side failed.

We are currently using **@learncard/core": "^7.0.1** Jackson suggested to upgrade, and we are looking forward to other ways to troubleshout this. 
Thanks for the help!

**Comment by Custard7** [comment]
@leonfhl - did upgrading to 8.0 @learncard/core resolve your issues?

https://docs.learncard.com/learn-card-sdk/learncard-core/migration-guide

**Comment by leonfhl** [comment]
Hi it didn't work, the upgrade worked to 8.0. But it didn't fix the issue. The output is the same.

**Comment by Custard7** [comment]
@leonfhl - are you able to post a sample of your code, or point to a running version of your issuer? Does DID auth work for other wallets, or are you just having issues with Veres?

Let's hop on a call in the am if you would like to troubleshoot live.

@TaylorBeeston any thoughts on what might be going wrong here?

**Comment by Custard7** [comment]
@leonfhl the storeCredentialViaChapiDidAuth is just a helper method, you can see it's full definition here:

https://github.com/learningeconomy/LearnCard/blob/main/packages/learn-card-core/src/wallet/plugins/chapi/chapi.ts#L55

You could directly call the methods on your learnCard instance and log the verification errors / log each step of the process to see where the issue is.

**Reply by leonfhl** [reply]
This helps, the method for us fail with Veres in:

![image](https://user-images.githubusercontent.com/5489406/200476347-7dfc94be-1b22-4dff-ad67-5ff9f60bd61c.png)

That was the exat step we skip with the manual DID Auth. So we are going to debug that part after Plugfest and keep you posted!

**Comment by Custard7** [comment]
@leonfhl - I was playing with the issuer you have at https://vc.acreditta.com:3007/

When I look at the DID in the issuer id of the credential payload:

<img width="1776" alt="Screen Shot 2022-11-07 at 10 27 31 PM" src="https://user-images.githubusercontent.com/2185016/200468185-f900c66d-d3ee-4d72-afd7-3ea07e557814.png">

`did:key:zz6MktiSzqF9kqwdU8VkdBKx56EYzXfpgnNPUAGznpicNiWfn`

This doesn't match the DID of your LearnBridge issuer, found at the following[ `/did` discovery endpoint](https://iaf5yytoj0.execute-api.us-east-1.amazonaws.com/did):

`did:key:z6Mkv1o2GEgtXjFdEMfLtupcKhGRydM8V7VHzii7Uh4aHoqH`

<img width="375" alt="Screen Shot 2022-11-07 at 10 27 58 PM" src="https://user-images.githubusercontent.com/2185016/200468249-72c8c219-c304-40cf-a4ee-bf0787a69e1e.png">

This will cause your issuance to fail! The credential must be issued from the DID of your issuer.

If you are using the `initLearnCard` with a vcApi url endpoint, you can specify the DID on construction:

```
// Initializes a LearnCard connected to a custom VC-API, with Issuer DID specified.
const customApi = await initLearnCard({ vcApi: 'https://iaf5yytoj0.execute-api.us-east-1.amazonaws.com', did: 'did:key:z6Mkv1o2GEgtXjFdEMfLtupcKhGRydM8V7VHzii7Uh4aHoqH' });

// Or, if you omit it, it will discover it automatically:
// Initializes a LearnCard connected to a custom VC-API that implements /did discovery endpoint.
const customApiWithDIDDiscovery = await initLearnCard({ vcApi: 'https://iaf5yytoj0.execute-api.us-east-1.amazonaws.com' });
```

https://docs.learncard.com/learn-card-sdk/learncard-core/plugins/official-plugins/vc-api#initialization

**Comment by Custard7** [comment]
@leonfhl if I swap out the issuer DID to the correct DID, I get a valid, signed credential in the response:

<img width="1867" alt="Screen Shot 2022-11-07 at 10 36 45 PM" src="https://user-images.githubusercontent.com/2185016/200469365-28ccddd1-d464-44b1-a9da-f3a765525761.png">


Also note, I went to validate your JSON of the credential you are signing, and I get the following:

https://www.jsonschemavalidator.net/s/5YExj6uV

<img width="1895" alt="Screen Shot 2022-11-07 at 10 38 58 PM" src="https://user-images.githubusercontent.com/2185016/200469630-f5d5bc8c-1862-41a7-9fa3-0dadec5fd9e4.png">

This was a common issue among plugfest participants, but the image field needs to be an object, like this:

https://www.jsonschemavalidator.net/s/koDOC4mv

<img width="963" alt="Screen Shot 2022-11-07 at 10 39 39 PM" src="https://user-images.githubusercontent.com/2185016/200469729-cb1de44b-874e-48da-8d8a-84c11c8c2f48.png">

**Comment by leonfhl** [comment]
Sorry that exact VC was a live attempt to put the JFF example directly in the DB, but we rapidly realized that didn't work so we erased and added manually from the ADD section so the issuer DID was inserted correctly. 

We managed to do the DID Auth but we have to do it manually and comment out the storeCredentialviaChapiDidAuth method:

![image](https://user-images.githubusercontent.com/5489406/200474034-a131ead2-35f3-40c5-9914-6495e102097a.png)
![image](https://user-images.githubusercontent.com/5489406/200474047-dd2cee80-e520-4ecf-868a-f7f8490ede3e.png)

We also found problems to put build instances in a server. Right now we are on [vc.acreditta.com](url) (without port) but the backend server is running pnpm development instance. We are going to report those later in other tread. 

Thanks for the time!

**Comment by Custard7** [comment]
@leonfhl great looks like this will work for now - and we can continue to iterate post-plugfest.

Let me know when we should make a video - is it ready?
